Applications/Spire/Source/Ui/DropDownMenu2.cpp

```diff
@@ -0,0 +1,195 @@
+  for(auto& item : items) {
+    m_list_layout->addWidget(item);
+    m_item_hovered_connections.AddConnection(item->connect_hovered_signal(
+      [=] { m_hovered_signal(item->get_value()); }));
+    m_item_selected_connections.AddConnection(item->connect_selected_signal(
+      [=] { on_item_selected(item->get_value(), item); }));
+  }
+  if(m_list_layout->count() > 0) {
+    update_height();
+    static_cast<DropDownMenuItem2*>(
+      m_list_layout->itemAt(0)->widget())->set_highlighted();
+    m_selected_index = 0;
+    m_current_index = 0;
+    m_selected_signal(get_value(*m_selected_index));
+    m_current_signal(get_value(*m_current_index));
+  }
+  hide();
+}
+
+connection DropDownMenu2::connect_current_signal(
+    const CurrentSignal::slot_type& slot) const {
+  return m_current_signal.connect(slot);
+}
+
+connection DropDownMenu2::connect_hovered_signal(
+    const HoveredSignal::slot_type& slot) const {
+  return m_hovered_signal.connect(slot);
+}
+
+connection DropDownMenu2::connect_selected_signal(
+    const SelectedSignal::slot_type& slot) const {
+  return m_selected_signal.connect(slot);
+}
+
+const QVariant& DropDownMenu2::get_value(int index) const {
+  if(m_list_layout->count() - 1 < index) {
+    static auto SENTINEL = QVariant();
+    return SENTINEL;
+  }
+  return get_item(index)->get_value();
+}
+
+boost::optional<int> DropDownMenu2::get_current() const {
+  return m_current_index;
+}
+
+void DropDownMenu2::set_current(int index) {
+  if(index >= get_count()) {
+    return;
+  }
+  if(m_current_index) {
+    get_item(*m_current_index)->reset_highlighted();
+  }
+  m_current_index = index;
+  auto highlighted_widget = get_item(*m_current_index);
+  highlighted_widget->set_highlighted();
+  scroll_to_current_index();
+  m_current_signal(highlighted_widget->get_value());
+}
+
+boost::optional<int> DropDownMenu2::get_selected() const {
+  return m_selected_index;
+}
+
+void DropDownMenu2::set_selected(int index) {
+  if(index < get_count()) {
+    on_item_selected(get_value(index), index);
+  }
+}
+
+int DropDownMenu2::get_count() const {
+  return m_list_layout->count();
+}
+
+DropDownMenuItem2* DropDownMenu2::get_item(int index) const {
+  return static_cast<DropDownMenuItem2*>(
+    m_list_layout->itemAt(index)->widget());
+}
+
+void DropDownMenu2::scroll_to_current_index() {
+  if(m_current_index != m_list_layout->count() - 1) {
+    m_scroll_area->ensureWidgetVisible(get_item(*m_current_index), 0, 0);
+  } else {
+    m_scroll_area->verticalScrollBar()->setValue(
+      m_scroll_area->verticalScrollBar()->maximum());
+  }
+}
+
+void DropDownMenu2::update_height() {
+  if(m_list_layout->count() == 0) {
+    return;
+  }
+  setFixedHeight(std::min(m_max_displayed_items, m_list_layout->count()) *
+    m_list_layout->itemAt(0)->widget()->height() + BORDER_PADDING());
+  m_scroll_area->setFixedHeight(height() - 1);
+}
+
+void DropDownMenu2::on_item_selected(const QVariant& value, int index) {
+  hide();
```

Comment:
Call ```hide``` after updating the index variables, structure mutations from those that are internal and not visible to those that are external and visible.